bundles/core/src/main/java/com/adobe/cq/wcm/core/components/commons/link/Link.java

```diff
@@ -0,0 +1,61 @@
+/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ ~ Copyright 2021 Adobe
+ ~
+ ~ Licensed under the Apache License, Version 2.0 (the "License");
+ ~ you may not use this file except in compliance with the License.
+ ~ You may obtain a copy of the License at
+ ~
+ ~     http://www.apache.org/licenses/LICENSE-2.0
+ ~
+ ~ Unless required by applicable law or agreed to in writing, software
+ ~ distributed under the License is distributed on an "AS IS" BASIS,
+ ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+package com.adobe.cq.wcm.core.components.commons.link;
+
+import java.util.Map;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.osgi.annotation.versioning.ConsumerType;
+
+import com.day.cq.wcm.api.Page;
+
+/**
+ * Describes a link target.
+ */
+@ConsumerType
+public interface Link {
+
+    /**
+     * Check if the link defined for the component is valid.
+     * @return true if component has a valid link defined
+     */
+    boolean isValid();
+
+    /**
+     * Externalized link URL.
+     * @return Link URL or null if link is invalid.
+     */
+    @Nullable
+    String getURL();
```

Comment:
I agree with @kwin . I suggested earlier to keep this method and to introduce a method that returns the resolved URL, based on the url string or on the target page:
```
public @Nullable String getResolvedURL()
```

Reply:
you're right: the javadoc is misleading here.

but keep in mind that this interface should represent *any* type of link, not only links to target pages. so it could also return a link to an external website, or to an asset (e.g. PDF download).
the goal of the method is to have a ready-to-use URL for all contexts. so it should be as mappend and as externalized as possible, and also usable in JSON output (where is no LinkTransformer running over it afterwards).
on the other hand we are not that far with the other parts for the implementation to really externalize URLs (no concept for site root #62, not means for configuring URLs to externalize for site context etc.). not possible to introduce all this in this PR we have to start simple.
what indeed could make sense is to introduce returning a fully mapped method here, we already have support for vanity paths.
i'm not a fan of using Sling Rewriter to rewrite/map URLs - my goal would be the core components are fully functional without any post processing transformation step. mapped URLs should work fine even a Sling Rewriter is running over it afterwards.

i would not like to introduce another URL like getResolvedURL, that get's very confusing for the developers using the interface - and will mostly use to misuse as no one understands the difference (think of ResourceResolver.getResource vs. ResourceResolver.resolve).

Reply:
The problem is that calling ResourceResolver.map() on an already mapped url leads to double URL encoding (https://issues.apache.org/jira/browse/SLING-1899). Therefore mapped urls which are processed by the Link Transformer might get invalid.

Reply:
i know, this flaw is in the Sling API for years - we have to de-encode it in the LinkHandler like [here](https://github.com/wcm-io/wcm-io-handler/blob/5a620b0fbbc1c99341f4bbd5978f5d6edcf80250/url/src/main/java/io/wcm/handler/url/impl/Externalizer.java#L54-L100). it's ugly, but works well; we're doing it for years.

Reply:
to correct myself: the code mentioned is not decoding the URL after calling the mapping methods. it should be kept URL-encoded if it contains special chars that are invalid unencoded in URLs. this way they should be valid in all use cases - HTL or JSON. i've not had a problem with sling rewriter. if you have a custom link rewriter in place you should decode the URL before processing it like [here](https://github.com/wcm-io/wcm-io-handler/blob/5a620b0fbbc1c99341f4bbd5978f5d6edcf80250/url/src/main/java/io/wcm/handler/url/rewriter/impl/UrlExternalizerTransformer.java#L101).

Reply:
but let keep us this PR small - it is a first step introducing a dedicated interfaces representing link, with not much more capabilites than the existing URL generation. let's add mapping support and all later. so maybe, just now only fix the misleading javadoc.

Reply:
For now I updated the Javadoc in (acf1edb) to mention that the URL has support for context-path and vanity paths. I'm not sure having something like `getOriginalURL` would be that helpful.

Reply:
IMHO it definitely needs to be maped (see e.g. https://github.com/adobe/aem-core-wcm-components/issues/426) for responses not handled by the link transformer, like JSON

Reply:
@kwin Would mapping as https://github.com/adobe/aem-core-wcm-components/pull/1411/commits/d7ed7a29c4c2ab96a86aa5f18802d61b2cbd4978#diff-3f692edefc2435524e017da0411675032dc8d070acf852ff46f291b73c146b39L182 be enough?

Reply:
Please always use the two argument version of map taking also the request into account. Some mapping rules may depend on the HOST header

Reply:
@kwin updated to two arguments version in 980b71e